doc/contribute: Add tip to run Checkpatch in pre-push hook

erwango opened the pull request:
Running Checkpatch in pre-commit hook could be a constraint
for day to day work. Though, running it before pushing to
zephyr repo.
Add a tip to run the script automatically before pushing thanks
to pre-psuh hook.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
DNM yet, does not work as expected in some cases...

erwango commented:
> DNM yet, does not work as expected in some cases...

False alert, work as expected

dbkinder commented on doc/contribute/index.rst:
```diff
@@ -362,6 +362,47 @@ it to contain:
+                fi
+
+                echo "Perform check patch"
+                exec git diff $range | /local/mcu/zephyr/zephyr-project/scripts/checkpatch.pl -
```
Hmmm, I suspect this path ``/local/mcu/zephyr/zephyr-project/scripts/checkpatch.pl`` should be ``${ZEPHYR_BASE}/scripts/checkpatch.pl``.   Has this been tested for Windows too?

erwango replied:
You're suspecting correctly, txs

marc-hb commented:
I think it would be better to recommend https://github.com/zephyrproject-rtos/ci-tools/blob/master/scripts/check_compliance.py which does a lot more and - unlike a shell script - should run on Windows too

erwango commented:
> Has this been tested for Windows too?

Nope. Status vs windows support will be same as existing proposed pre-commit. Let's say this proposal is iso-support vs existing doc.

> I think it would be better to recommend https://github.com/zephyrproject-rtos/ci-tools/blob/master/scripts/check_compliance.py which does a lot more and - unlike a shell script - should run on Windows too

This requires ci-tools installation, which is nowhere mentioned in zephyr main repo current documentation. So I think a pre-requisit for this proposal (which is indeed valuable) would be an introduction to ci-tool within zephyr doc.

marc-hb commented:
> This requires ci-tools installation

As far as I remember ci-tools "installation" was merely cloning zephyr-ci-tools and running `pip3 -r requirements` so very little extra compared to the initial zephyr install.

I really don't see value in this incomplete duplication of python code from zephyr-ci-tools into shell script code in the main repo. Assuming there were value, at least that shell script should be in an actual `run-checkpatch.sh` file that people can run and maintain as opposed to having to copy/paste the documentation.

-1 from me.

ulfalizer commented:
> > This requires ci-tools installation
> 
> As far as I remember ci-tools "installation" was merely cloning zephyr-ci-tools and running `pip3 -r requirements` so very little extra compared to the initial zephyr install.
> 
> I really don't see value in this incomplete duplication of python code from zephyr-ci-tools into shell script code in the main repo. Assuming there were value, at least that shell script should be in an actual `run-checkpatch.sh` file that people can run and maintain as opposed to having to copy/paste the documentation.
> 
> -1 from me.

I think we should move `ci-tools` back into the Zephyr repo (it used to be part of it). It gets awkward to have it as a separate repository when some CI check needs to be updated in tandem with a change in the main repo, and it makes it harder for people to find the code too.

If `ci-tools` stays as a separate repo, I'm all for hinting people towards `scripts/check_compliance.py` at least. It's hard to discover now.

marc-hb commented:
At the risk of stating the obvious, @ulfalizer, @nashif  and myself are actively making check_compliance.py better right now: https://github.com/zephyrproject-rtos/ci-tools/pulls
Please join us.

marc-hb reviewed (dismissed):
At least do this which provides per commit feedback. Besides being less specific, checking everything as one giant patch could also miss stuff:

`${ZEPHYR_BASE}/scripts/checkpatch.pl --git $range`

Please also mention check_compliance.py as a more complete alternative.

marc-hb commented:
> at least that shell script should be in an actual `checkpatch-series-push-hook.sh` file that people can run and maintain as opposed to having to copy/paste the documentation.

This would also let Linux users "install it" with a mere symlink and gets updates without doing anything.

erwango commented:
@marc-hb, modified with a hook callable/maintainable script.

erwango committed:
scripts: Add hook calllable series-push-hook script

Running checkpatch in pre-commit hook could be a constraint
for day to day work. Though, running it before pushing to
zephyr repo is highly recommended.
Add a  series push script callable from pre-push hook.
Document how to use it in contribute section.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

marc-hb reviewed (commented):
Thanks!

dbkinder reviewed (approved):
LGTM, thanks.
